Reduce by 2 the memory requirement in `generate()` 🔥🔥🔥

Cyrilvallez opened the pull request:
# What does this PR do?

Change the data structure and implementation of `DynamicCache` to **halve** the memory requirement in `generate()` at no noticeable speed degradation.

## Reason

I was working on precise memory estimation of `generate()` and noticed that the expected memory peak and the one observed are different (observed one is much higher). I was able to track down the problem to the current implementation of `DynamicCache`.

Since `torch.cat()` creates a copy, and that during `generate` we loop over each token using the previous cache in the inputs (thus the old cache is referenced in the inputs so cannot be garbage collected), the current implementation basically has 2 copies of the **full** cache in every iteration. By changing the data structure from `Tuple[Tuple[Tensor]]` to `Tuple[Tuple[List[Tensor]]]`, I was able to avoid this copy and reduce the memory footprint of `generate()` by 2. 

However, that means that potentially thousands of tensors must be `cat()`'ed when `cache.update()` is called to feed the correct tensor to the `AttentionLayer`. This results in a speed degradation. I was able to mitigate this issue by periodically `cat()`'ing the tensors in the cache when there are more than `N` (I used 50 as of now), which incurs a negligible memory increase, as the cache for a sequence length of `N=50` is usually completely negligible compared to the size of the full cache (hundreds, or even more frequently thousands). This strategy almost completely removes the speeds degradation, allowing to get the best of both worlds. `N` could even be chosen dynamically in `generate()` depending on the input length and max new tokens, but 50 seemed like a good heuristic to start with.

Basically, at a very small performance penalty that is visible only for very large sequence length, we reduce the memory footprint by 2, which by itself allows to increase the batch size by 2 (so should be able to actually speed up the process as passing a sequence of 2 times the batch at lower speed should still be faster then runnings 2 loops at faster speed).

Of course, the best would still be to use a `StaticCache` as I saw that you started implementing, but a DynamicCache is still very much useful, and should not imply to double its effective memory footprint when used in loop referencing itself.

## Benchmark

Here you can see the benchmark I ran using a RTX 4090 (24 GB) and Mistral 7B, Llama2 7B and Llama3 8B.

### Fix batch size of 1 and input length of 300, variable new token number

![Mistral-7B-v0.1_memory_fix_batch.pdf](https://github.com/huggingface/transformers/files/15149403/Mistral-7B-v0.1_memory_fix_batch.pdf)
![Mistral-7B-v0.1_time_fix_batch.pdf](https://github.com/huggingface/transformers/files/15149405/Mistral-7B-v0.1_time_fix_batch.pdf)
![Llama-2-7b-hf_memory_fix_batch.pdf](https://github.com/huggingface/transformers/files/15149408/Llama-2-7b-hf_memory_fix_batch.pdf)
![Llama-2-7b-hf_time_fix_batch.pdf](https://github.com/huggingface/transformers/files/15149410/Llama-2-7b-hf_time_fix_batch.pdf)
![Meta-Llama-3-8B_memory_fix_batch.pdf](https://github.com/huggingface/transformers/files/15149411/Meta-Llama-3-8B_memory_fix_batch.pdf)
![Meta-Llama-3-8B_time_fix_batch.pdf](https://github.com/huggingface/transformers/files/15149412/Meta-Llama-3-8B_time_fix_batch.pdf)

### Fix new token number of 2000 and input length of 300, variable batch size

![Mistral-7B-v0.1_memory_fix_length.pdf](https://github.com/huggingface/transformers/files/15149416/Mistral-7B-v0.1_memory_fix_length.pdf)
![Mistral-7B-v0.1_time_fix_length.pdf](https://github.com/huggingface/transformers/files/15149418/Mistral-7B-v0.1_time_fix_length.pdf)
![Llama-2-7b-hf_memory_fix_length.pdf](https://github.com/huggingface/transformers/files/15149419/Llama-2-7b-hf_memory_fix_length.pdf)
![Llama-2-7b-hf_time_fix_length.pdf](https://github.com/huggingface/transformers/files/15149420/Llama-2-7b-hf_time_fix_length.pdf)
![Meta-Llama-3-8B_memory_fix_length.pdf](https://github.com/huggingface/transformers/files/15149421/Meta-Llama-3-8B_memory_fix_length.pdf)
![Meta-Llama-3-8B_time_fix_length.pdf](https://github.com/huggingface/transformers/files/15149422/Meta-Llama-3-8B_time_fix_length.pdf)

## Integration in Transformers

As I am changing the data structure of the DynamicCache, I had to modify how it is used in models modeling. For now, I only modified `LlamaForCausalLM` and `MistralForCausalLM` which are using the `DynamicCache` by default to test my implementation. Also, the change of data structure may have impacts elsewhere that I overlooked (if other code rely on attributes of the cache, e.g. a call to `cache[0][0].shape` would be an `AttributeError` now). If so, do not hesitate to point me towards these parts and I can modify it.

I would be happy to help integrate the change in all models if you decide to move forward with the PR.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?


## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

<!-- Your PR will be replied to more quickly if you can figure out the right person to tag with @

 If you know how to use git blame, that is the easiest way, otherwise, here is a rough guide of **who to tag**.
 Please tag fewer than 3 people.

Models:

- text models: @ArthurZucker and @younesbelkada
- vision models: @amyeroberts
- speech models: @sanchit-gandhi
- graph models: @clefourrier

Library:

- flax: @sanchit-gandhi
- generate: @gante
- pipelines: @Narsil
- tensorflow: @gante and @Rocketknight1
- tokenizers: @ArthurZucker
- trainer: @muellerzr and @pacman100

Integrations:

- deepspeed: HF Trainer/Accelerate: @pacman100
- ray/raytune: @richardliaw, @amogkam
- Big Model Inference: @SunMarc
- quantization (bitsandbytes, autogpt): @SunMarc and @younesbelkada

Documentation: @stevhliu and @MKhalusova

HF projects:

- accelerate: [different repo](https://github.com/huggingface/accelerate)
- datasets: [different repo](https://github.com/huggingface/datasets)
- diffusers: [different repo](https://github.com/huggingface/diffusers)
- rust tokenizers: [different repo](https://github.com/huggingface/tokenizers)

Maintained examples (not research project or legacy):

- Flax: @sanchit-gandhi
- PyTorch: See Models above and tag the person corresponding to the modality of the example.
- TensorFlow: @Rocketknight1

 -->

@ArthurZucker and @younesbelkada and @gante

ArthurZucker reviewed (commented):
Great PR and great in depth analysis. 
One thing we could / should also do is remove this reference no? 
For now we pass the cache as an input to every single layer, which also returns it. 
I suppose that having less of this could solve what you mention about having 2 copies of the cache. 

Anyhow if we can fix the memory consumption let's go 🔥  
Changing the cache format is super breaking and cannot be done easily but we could do our best to avoid copies!

Cyrilvallez commented:
Hi @ArthurZucker, thanks for the feedback!

Unfortunately, I don't think we can handle all cases by keeping the same data structure. For example, if someone passes the `past_key_values` to `generate()` to resume generation, there is no way to avoid having the reference around as it will be referenced outside "transformer's scope" (e.g. the global scope). In this case, the current data structure will always create a copy and return the copy while still having the old one in memory because we cannot just dynamically add more memory to an already existing tensor. With the new data structure however, we can simply append to the old cache in-place. I just tested it and it works nicely, without allocating any more memory (except for the memory needed for adding a few tokens to the already existing cache of course)!

I checked the rest of the code base and I agree that it is a big change, but it is very much worth it. I'm definitely down to collaborate on it if you wish!

Cyrilvallez committed:
Fix contrastive_search for new cache structure, and improve performance by removing inneficient torch.stack(torch.split(x, top_k, dim=0))

Cyrilvallez committed:
Fix _contrastive_search for non-standard cache using ellipsis slicing

Cyrilvallez commented:
Hi @ArthurZucker,

I am done with the work. Could you please review it? At this point, it should be 100% backward compatible. The only change is that now `past_key_values` will be returned as a `Tuple[Tuple[List[Tensor]]]`, and be modified in-place if directly supplied to `model.generate()` for any of the model architectures that use `DynamicCache` by default (i.e. Cohere, Dbrx, Gemma, Idefics2, Llama, Mistral, Mixtral, Olmo, Persimmon, Phi, Phi3, Qwen2, Qwen2Moe, StableLM, StarCoder2 as of now). 
Every other model will still use the old format `Tuple[Tuple[Tensor]]` (or their own custom format, e.g. for Bloom) but will NOT benefit from the improved memory performance.

Here is a final benchmark of the improved performances for every decoding strategy available in Transformers. It was still done on one of my RTX 4090, and with Mistral-7B-v0.1:

![contrastive_search.pdf](https://github.com/huggingface/transformers/files/15218488/contrastive_search.pdf)
![greedy.pdf](https://github.com/huggingface/transformers/files/15218490/greedy.pdf)
![sample.pdf](https://github.com/huggingface/transformers/files/15218492/sample.pdf)
![beam_search.pdf](https://github.com/huggingface/transformers/files/15218493/beam_search.pdf)
![beam_sample.pdf](https://github.com/huggingface/transformers/files/15218494/beam_sample.pdf)
![constrained_beam_search.pdf](https://github.com/huggingface/transformers/files/15218495/constrained_beam_search.pdf)
![group_beam_search.pdf](https://github.com/huggingface/transformers/files/15218496/group_beam_search.pdf)
![assisted.pdf](https://github.com/huggingface/transformers/files/15218498/assisted.pdf)

Note that for assisted decoding, for 1000 tokens and more I run into a weird CUDA error (`[/opt/conda/conda-bld/pytorch_1704987289929/work/aten/src/ATen/native/cuda/Indexing.cu:1237]: indexSelectSmallIndex: block: [29,0,0], thread: [84,0,0] Assertion 'srcIndex < srcSelectDimSize' failed.` and `RuntimeError: CUDA error: device-side assert triggered`) in BOTH the actual Transformers version and my modified script, which is why this particular benchmark stops after 500 tokens. I don't know exactly what is happening here, but it is not due to my changes as it is also happening in the current version.

I hope this can be integrated as soon as possible! Don't hesitate to tell me if I am missing something that should be done before merging.

Cyrilvallez committed:
Fix all outputs.logits memory leaks for all decoding strategies!

Cyrilvallez committed:
Fix small error in _contrastive_search()

Cyrilvallez commented:
Hi @ArthurZucker,

I was investigating why we observe those "minimal improvements" even with very large input sizes and just 2 new tokens. I found out that the reason was that the logits in every iteration leaked to the next iteration though some hanging references to the outputs dictionary. This was true for every decoding strategy except assisted decoding. I fixed it in my last 2 commits.

## Non-beam methods

That means that for any input size, the memory peak was not caused by the forward pass of the first iteration with the large input, but during the 2nd iteration, when the copying of the full cache was happening. At that point, the memory peak consisted of `2*past_KV + first_iteration_logits` (the logits in the first iteration when all tokens are feeded to the model can quickly grow very large). After my first modifications, we were no longer keeping the 2nd cache copy, thus in the second iteration, the peak would be `past_KV + first_iteration_logits`, which is always at least smaller or equal to the memory needed for the forward in the first iteration (this pass is creating those values so need at least that much memory, in practice more). Thus the leaking of logits was not a memory bottleneck after solving the copying of the cache in this case. I still solved it though, because it does not hurt performances and would be weird to keep. In beam methods however, removing it is a big improvement (see below).
Anyway, that means that we can compute the "minimal improvement" as:

```
minimal_improvement = (2*past_KV + first_iteration_logits) / memory(first_forward_pass)
```

Now, both the cache and the logits trivially scale linearly with the input size. Since flash attention, the memory needed for the forward pass also scales linearly with the input size, thus `minimal_improvement` is a constant for all input sizes. This is exactly what I observed in my past comment, with this constant being ~1.22 for Mistral and ~1.62 for Llama2. I computed the ratio presented above and obtained the same values, which confirmed my analysis. 

Now, as we generate more tokens after the first iteration, and the cache size exceeds the `memory(first_forward_pass)`, we tend to the `maximal_improvement = 2`, because we removed the perpetual copy of the cache (logits are negligible here because after the first pass, the logits will always be for only 1 token, thus small in size). 

## Beam methods

For beam methods, the story is slightly different. Because reordering the cache allocates a full copy of the cache, the leak of the cache and the logits caused `3*past_KV + first_iteration_logits` to be allocated in memory during the 2nd iteration. After my work, it is now only `2*past_KV` (I believe we could even reorder in a smarter way that would avoid that 2nd copy, but that would be future work). That means, as before, the `minimal_improvement` is:

```
minimal_improvement = (3*past_KV + first_iteration_logits) / memory(first_forward_pass)
```

and when generating enough new tokens, the `maximal_improvement` is going from `3*past_KV` to `2*past_KV`, thus 1.5. However, in this case, `minimal_improvement` is usually always bigger than 1.5, meaning that if we generate few tokens compared to the size of the input we will benefit from memory improvement bigger than 1.5, and if we generate enough new tokens compared to the input size, the improvement will tend to 1.5.

ArthurZucker commented on src/transformers/generation/utils.py:
cc @gante super interesting commit!

ArthurZucker commented on src/transformers/cache_utils.py:
```diff
@@ -136,42 +136,72 @@ def update(
+            self.key_cache[layer_idx].append(key_states)
+            self.value_cache[layer_idx].append(value_states)
+
+        # Whenever we have more than N new K-V value, cat() them. That way, we keep a relatively low number
+        # of tensors in self.key_cache[layer_idx], which is more efficient to later cat() them all, and we only
+        # copy a small subset into memory whenever we cat() the last N K-V states
+        N = 50
+        index = None
+        for i, x in enumerate(self.key_cache[layer_idx]):
+            if x.shape[-2] == 1:
+                index = i
+                break
+        if index is not None and len(self.key_cache[layer_idx]) - 1 - index > N:
+            self.key_cache[layer_idx] = self.key_cache[layer_idx][:index] + [
+                torch.cat(self.key_cache[layer_idx][index:], dim=-2)
+            ]
+            self.value_cache[layer_idx] = self.value_cache[layer_idx][:index] + [
+                torch.cat(self.value_cache[layer_idx][index:], dim=-2)
+            ]
```
an interesting way to do some bucketing

ArthurZucker commented on src/transformers/cache_utils.py:
```diff
@@ -136,42 +136,72 @@ def update(
+        # Whenever we have more than N new K-V value, cat() them. That way, we keep a relatively low number
+        # of tensors in self.key_cache[layer_idx], which is more efficient to later cat() them all, and we only
+        # copy a small subset into memory whenever we cat() the last N K-V states
+        N = 50
```
can be a generation_config argument

ArthurZucker commented on src/transformers/generation/utils.py:
```diff
@@ -1302,7 +1302,10 @@ def _get_initial_cache_position(self, input_ids, model_kwargs):
             else:
-                past_length = model_kwargs["past_key_values"][0][0].shape[2]
+                if isinstance(model_kwargs["past_key_values"][0][0], list):
+                    past_length = sum(x.shape[-2] for x in model_kwargs["past_key_values"][0][0])
```
if the past key values is a EfficientCache class, this should be easier to extract. + @gante using the cache_positions would be the best here!

ArthurZucker reviewed (commented):
@Cyrilvallez kudos for this really high quality contribution 🤗 🚀 
Your in-depth explanations are very useful, and all of it makes sense. 
I think at some point we did want to change the way the cache was stored, to also remove the two transpose operation that are always needed in the modeling code (which is not the case for jax codes). 

@gante some potential break is the cache format, but I think we can add an `EfficientDynamicCache` class that would be used upon activation. @Cyrilvallez this would be for use better than having two code paths and makes more sense for isolating bugs / maintenance! 

PS: do you have any gist on how to reproduce memory benchmarks and etc?

WDYT @gante

Cyrilvallez commented:
> @Cyrilvallez kudos for this really high quality contribution 🤗 🚀 Your in-depth explanations are very useful, and all of it makes sense. I think at some point we did want to change the way the cache was stored, to also remove the two transpose operation that are always needed in the modeling code (which is not the case for jax codes).
> 
> @gante some potential break is the cache format, but I think we can add an `EfficientDynamicCache` class that would be used upon activation. @Cyrilvallez this would be for use better than having two code paths and makes more sense for isolating bugs / maintenance!
> 
> PS: do you have any gist on how to reproduce memory benchmarks and etc?
> 
> WDYT @gante

Thanks for reviewing @ArthurZucker! Here is a link to the benchmarks I ran: https://gist.github.com/Cyrilvallez/ce1adfad1d561c1e8dc92666ab5a9e8c

It is a bit messy but you should find everything. Each benchmark was run on the official version 4.40.1, saving outputs with `legacy` in the filenames, and then on my current branch using `patched` in the filenames. You should manually change the filenames depending on which version you use if you re-run the benchmarks. Also, you may need to modify the actual paths to those filenames when creating the figures. Apart from that, I think the link should contain everything you need to reproduce.

ArthurZucker commented:
@Cyrilvallez what do you think about creating a new cache class?

Cyrilvallez commented:
@ArthurZucker I agree that a new class would be clearer and more maintainable. Not sure what you meant by "that would be used upon activation" however? I would use the new EfficientDynamicCache as the new default for all models actually using DynamicCache by default if not provided. As I cannot think of any backward compatibility issues (even if some people save cache to file to restart from it, the new class will work), doing otherwise would be counter-intuitive I think (it would be weird to have to pass an argument to activate a feature that does the same as before, but more efficiently in all cases). I don't have your experience maintaining such a widely used codebase however, so if you think people may run into problems, we could raise a warning to point them to how to use the old DynamicCache?

Cyrilvallez commented:
Ok! I will make the necessary changes next Monday 👌🏻

gante commented:
@Cyrilvallez very cool in-depth exploration 😮🔥 And also very impactful consequences of the suggested changes!

Reading the discussion, from a usage perspective, I agree with Arthur: a separate class would be ideal! With a separate cache class, we:
1. Can keep a simple reference cache class in `DynamicCache`;
2. The default class used and returned in `forward` stays as simple as possible;
3. [after a proper deprecation cycle] Can use the new efficient class by default in `generate`, getting all the benefits showcased in the benchmarks above.

I'll review the full code after it is implemented as a separate class, as we all seem aligned on what should be done 🤗 @Cyrilvallez ping me when it's ready :D

gante commented:
@Cyrilvallez out of curiosity: have you considered/explored expanding the cache with fixed-size blocks every time we hit the limit, similarly to paged attention?

Cyrilvallez committed:
Make all necessary change and revert for the new class

Cyrilvallez committed:
Apply coding style

Cyrilvallez committed:
Remove pipes in type hints for compatibility

Cyrilvallez committed:
correct type hint

Cyrilvallez committed:
apply style

Cyrilvallez commented:
@ArthurZucker @gante I realized yesterday that what actually creates the copies is not the current `DynamicCache` itself, but the back and forth `from_legacy_cache` and `to_legacy_cache` calls (that creates tuples that are immutables). Thus, forcing to use proper `DynamicCache` in `generate()` before going into specialized decoding strategy would retain all the above benefits while removing the need for a new class. Something like:

```
if use_cache and model_supports_dynamic_cache_class:
    if generation_config.cache_implementation is None and not isinstance(model_kwargs.get('past_key_values', None), Cache):
        model_kwargs['past_key_values'] = DynamicCache()
```

Doing so will avoid the back and forth switch from legacy to cache, and thus all references in the decoding strategies functions will point to the same `Cache` object that will be modified in-place for all of them because the caches are List[Tensor]. 

This would greatly simplify the above approach while retaining all benefits (even more as `reorder` would also happen in-place layer after layer, thus not allocating new memory -> beam decoding will use 3x less memory instead of 1.5). The small speed decrease would also be removed.

Cyrilvallez committed:
Use DynamicCache by default and solve conflicts

Cyrilvallez committed:
Fix rebase issues

Cyrilvallez committed:
Add `_supports_dynamic_cache_class` in models for models that support DynamicCache but not other caches to make DynamicCache the default for more models

Cyrilvallez committed:
Create generation config to return legacy format by default, or to choose not to

Cyrilvallez committed:
style

Cyrilvallez committed:
Fix case when use_cache is False

Cyrilvallez committed:
Remove default DynamicCache in assiste_decoding if assistant_model does not support it + fix _seen_tokens when cropping cache

Cyrilvallez committed:
Update prepare_inputs_for_generation() for case with empty DynamicCache

Cyrilvallez committed:
Correct return of args in _assisted_decoding

Cyrilvallez committed:
Remove EfficientDynamicCache as it is no longer needed

Cyrilvallez commented:
@ArthurZucker @gante The work is ready for final review! 

As previously said, `EfficientDynamicCache` was not needed in the end. I rebased my branch to remove previous commits and make everything clearer. Models with support for `DynamicCache` will use it from start to end as of now, without back and forth to the old legacy format. However, `generate()` will return the legacy format by default for backward compatibility. This can be controlled with the generation config argument `return_legacy_cache` if this is not wanted (e.g. in chat applications, when restarting from previous cache, it is more efficient to always keep a `DynamicCache` to avoid any copying and get between 3X and 2X memory improvement even for sample and greedy decoding as previously mentionned).

Models without support for `DynamicCache` will still benefit from improvements as well as I removed all the references to leaked logits/leaked copies of past_key_values. 
Moreover, this paves the way to the full adoption of the new `Cache` format everywhere as all generate functions can now accomodate a `DynamicCache` instance all the way.

Finally, here is the final be
![beam-based.pdf](https://github.com/huggingface/transformers/files/15324203/beam-based.pdf)
nchmark of performances:
![non_beam-based.pdf](https://github.com/huggingface/transformers/files/15324206/non_beam-based.pdf)
![contrastive.pdf](https://github.com/huggingface/transformers/files/15324207/contrastive.pdf)
![contrastive_top_k.pdf](https://github.com/huggingface/transformers/files/15324208/contrastive_top_k.pdf)

Beam-based methods now benefit from using 3X less memory because even `reorder()` is made in-place. Non beam-based are the same as before (except contrastive search). Contrastive benefits from 2X less memory as `top_k` increases , and even more for small `top_k` because expanding the cache is also done more efficiently now.

Finally, as we are using `DynamicCache`, we ne longer have any speed degradation. All those memory improvements are FREE LUNCH all the way.

Cyrilvallez commented:
New idea to further improve memory: #30860

ArthurZucker commented:
Damn that's impressive! Reviewing now!

ArthurZucker commented on src/transformers/generation/candidate_generator.py:
```diff
@@ -387,10 +387,7 @@ def _crop_past_key_values(model, past_key_values, maximum_length):
-            if past_key_values.value_cache[idx].shape[-1] != 0:
-                past_key_values.key_cache[idx] = past_key_values.key_cache[idx][:, :, :maximum_length, :]
-                past_key_values.value_cache[idx] = past_key_values.value_cache[idx][:, :, :maximum_length, :]
+        past_key_values.crop(maximum_length)
```
this is super nice

ArthurZucker commented on src/transformers/generation/configuration_utils.py:
```diff
@@ -382,6 +384,7 @@ def __init__(self, **kwargs):
 
         # Cache implementation
         self.cache_implementation = kwargs.pop("cache_implementation", None)
+        self.return_legacy_cache = kwargs.pop("cache_implementation", True)
```
bit weird to fetch `cache_implementation`

Cyrilvallez replied:
Indeed, should be `return_legacy_cache`, this one is clearly my mistake!

ArthurZucker commented on src/transformers/generation/utils.py:
```diff
@@ -1993,14 +2027,17 @@ def _contrastive_search(
 
                 # next logit for contrastive search to select top-k candidate tokens
-                logit_for_next_step = outputs.logits[:, -1, :]
+                # Clone is needed to avoid keeping a hanging ref to outputs.logits which may be very large for this first iteration
+                # (the clone itself is always small)
+                logit_for_next_step = outputs.logits[:, -1, :].clone()
```
💘

ArthurZucker commented on src/transformers/generation/utils.py:
```diff
@@ -3698,6 +3802,9 @@ def _split(data, full_batch_size: int, split_size: int = None):
         return [None] * (full_batch_size // split_size)
     if isinstance(data, torch.Tensor):
         return [data[i : i + split_size] for i in range(0, full_batch_size, split_size)]
+    # New efficient cache
+    elif isinstance(data, DynamicCache):
+        return data.split(full_batch_size, split_size)
```
that's a nice way to put it IMO

ArthurZucker commented on src/transformers/models/cohere/modeling_cohere.py:
```diff
@@ -1165,7 +1166,11 @@ def prepare_inputs_for_generation(
     ):
         past_length = 0
-        if past_key_values is not None:
+        # We may have an initialized but empty DynamicCache during first iteration
+        past_exist = past_key_values is not None and not (
+            isinstance(past_key_values, DynamicCache) and len(past_key_values) == 0
+        )
+        if past_exist:
```
if we are in generate, this should not even exist anyway and cache positions should be used!

ArthurZucker replied:
this particular piece of code is a blocker for me. It's something we should and can handle in generate no?

ArthurZucker reviewed (commented):
Alright! In general 🔥 looks great
- I think splitting the PR into 3 would be the best!
1. PR with the clone and del of the output
2. PR with the removal of the support for tuples 
3. PR with the new efficient cache. 
Unless 2 and 3 have to go together! 

@gante will be the one reviewing, and deciding so let's get his opinion! 🤗 

In any case, amazing work! 🚀

gante commented on src/transformers/generation/utils.py:
```diff
@@ -1102,7 +1102,13 @@ def _validate_model_class(self):
         # If a `Cache` instance is passed, checks whether the model is compatible with it
-        if isinstance(model_kwargs.get("past_key_values", None), Cache) and not self._supports_cache_class:
+        past = model_kwargs.get("past_key_values", None)
+        if isinstance(past, DynamicCache) and not self._supports_dynamic_cache_class:
+            raise ValueError(
+                f"{self.__class__.__name__} does not support an instance of `DynamicCache` as `past_key_values`. Please "
+                "check the model documentation for supported cache formats."
+            )
```
I believe this new `if` and  the `_supports_dynamic_cache_class` model attribute are redundant: all models with `self._supports_cache_class = True` should support `DynamicCache`. As such, both this `if` and the new model attribute can be removed.

The exception is Jamba, which has a custom `Cache` (and no support for the legacy class). lmk if custom logic is needed for Jamba, so we can find a solution that doesn't require a new model attribute 🤗

Cyrilvallez replied:
Ha yes, this has indeed become redundant since the clarification for cache supports attributes in https://github.com/huggingface/transformers/commit/9d889f870ed202de2859b0236895fe7f985a0c72! No need for this attribute anymore then 👌 I rebased my branch to apply latest changes in main, and remove the attribute

gante commented on src/transformers/generation/utils.py:
```diff
@@ -1671,14 +1688,23 @@ def generate(
             if generation_config.cache_implementation == "static":
                 raise ValueError("assisted generate is not supported with `static_cache`")
 
+            # Remove potential default DynamicCache if assistant does not support it
+            assistant_kwargs = copy.copy(model_kwargs)
+            if assistant_model is not None:
+                if use_dynamic_cache_by_default and not assistant_model._supports_dynamic_cache_class:
+                    if len(assistant_kwargs["past_key_values"]) == 0:
+                        del assistant_kwargs["past_key_values"]
+                    else:
+                        assistant_kwargs["past_key_values"] = assistant_kwargs["past_key_values"].to_legacy_cache()
```
request: can we move this logic to `AssistedCandidateGenerator.__init__`? That way, all logic regarding the initialization of the assistant attributes stays in one place :)

Cyrilvallez replied:
Yes! This will indeed be much cleaner

gante reviewed (commented):
Very very cool developments! It's amazing to see that we could get the memory benefits without adding a new class 💛 Really good work, @Cyrilvallez 🔥 

I've added a few comments, mostly related to long-term maintenance 

@ArthurZucker @Cyrilvallez -- I'm happy with this being done in a single PR :) After we 3 are happy with the changes, before merging, I will run a few slow tests on my end to confirm the PR is fully backwards compatible!

Cyrilvallez committed:
Correct mistake in generation config

Cyrilvallez committed:
Move cache logic of assisted decoding to AssistedCandidateGenerator.__init__

Cyrilvallez committed:
change DynamicCache function names from "split" to "batch_split" for readability + apply coding style

Cyrilvallez committed:
Remove `_supports_dynamic_cache_class` attribute after rebase

Cyrilvallez committed:
Correct missing line lost in conflict resolution during rebasing

Cyrilvallez committed:
Add special case for Jamba

Cyrilvallez committed:
Fix jamba test

Cyrilvallez committed:
Coding style

Cyrilvallez commented:
@ArthurZucker @gante I applied all changes following your comments! 

Repo consistency and code quality errors do not come from files I modified (I think code quality errors come from ruff version now being 0.4.4, previously 0.1.x).
Same goes for tests_tf, I did not touch any tf files either.

gante reviewed (approved):
LGTM, thank you for iterating and making `transformers` better for all of us 💛

gante commented:
@Cyrilvallez I believe CI is red due to external causes, [this PR](https://github.com/huggingface/transformers/pull/30932) should fix it after it gets merged

@ArthurZucker can you give this PR a final check? :)

gante commented:
Ran a few tests locally (slow tests for popular models, slow `generate` tests, slow cache tests), everything is passing 👍

ArthurZucker commented:
Having a look!

ArthurZucker commented:
I wanted to run the slow tests but did not work 😓

Cyrilvallez commented:
@ArthurZucker Should I try to rebase once again to see if it solves the issue?

younesbelkada commented:
@Cyrilvallez please yes, rebasing should work !

Cyrilvallez committed:
coding style

Cyrilvallez committed:
Correct missing import in rebasing

Cyrilvallez committed:
Simplify _validate_model_kwargs based on removal of _supports_dynamic_cache attribute

Cyrilvallez committed:
Simplify code paths in _contrastive_search

Cyrilvallez committed:
coding style

Cyrilvallez commented:
Hi @zucchini-nlp! When rebasing I noticed that in your recent https://github.com/huggingface/transformers/pull/30483 you made QuantizedCache a subclass of DynamicCache. However, some code paths in `generate()` rely on checking `isinstance(cache, DynamicCache` (this was the case before this PR in `_contrastive_search`). This is likely to crash if a Quantizedcache instance go inside one of those paths. 

Before, I had added a few methods (`split`, `batch_split`, `from_batch_split` to make the code go through cache methods instead of specialized snippets elsewhere. After I saw your PR, I added `batch_repeat_interleave` and `batch_select_indices` in `DynamicCache` to use in `_contrastive_search` instead of hard-coding what should happen inside this function. At this point, I think that there are no more code paths left that hard-code a specific behavior that do not use a `DynamicCache` method.

This means that if `QuantizedCache` overwrites these 5 methods (I think it is likely to be needed, but did not look at `QuantizedCache` in details maybe default implementation in `DynamicCache` is enough?) , everything will work correctly in all current code paths checking `isinstance(cache, DynamicCache)` in `generate`. As of now, I think that at least contrastive search is sure to fail in e.g. this part https://github.com/huggingface/transformers/blob/main/src/transformers/generation/utils.py#L2182

If this PR gets merged, overwriting the 5 methods mentioned above should make `QuantizedCache` work in any scenario.

Cyrilvallez commented:
@ArthurZucker Rebasing is done and all CIs are green!

zucchini-nlp commented:
@Cyrilvallez Right, QuantizedCache stores most of the past kv in a private list, so I think these methods would not work even before your changes. 

Thanks for noticing! I will think if we want to make quantized cache compatible with these generation strategies, otherwise raise an error when user tries to go that path. 

And amazing work in this PR for making generation better! 🔥

Cyrilvallez commented:
Yes, it will fail in the current state of the library, but adding support will be straight-forward and won't require messing with 'generation.utils' after this is merged!

ArthurZucker commented:
Reviewing today sorry for the delay

Cyrilvallez commented:
Hi @ArthurZucker, any news? 😁

ArthurZucker commented on src/transformers/cache_utils.py:
```diff
@@ -389,6 +389,57 @@ def from_legacy_cache(cls, past_key_values: Optional[Tuple[Tuple[torch.FloatTens
                 cache.update(key_states, value_states, layer_idx)
         return cache
 
+    def crop(self, maximum_length: int):
```
We need to update the docstring of the class to explain why we have these methods! 🤗

ArthurZucker replied:
```python
class DynamicCache(Cache):
    """
    A cache that grows dynamically as more tokens are generated. This is the default for generative models.
    It stores the Key and Value states as a list of tensors, one for each layer. The expected shape for each tensor is
    `[batch_size, num_heads, seq_len, head_dim]`.
    """
```
this needs to be updated!

ArthurZucker reviewed (commented):
LGTM appart from the modeling modifications!

Cyrilvallez committed:
Update docstrings of cache methods

Cyrilvallez committed:
Update prepare_inputs_for_generation() -> past_key_values are always Cache objects

Cyrilvallez commented:
@ArthurZucker The modeling update was mostly used in the case `inputs_embeds` were passed, but checking `past_length == 0` is equivalent and cleaner so I updated. I also removed the `TODO: remove if-else check after generate prioritize Cache objects` as this is now the case.

Should be good now!

ArthurZucker reviewed (approved):
Okay! 🚀 a last nit and let's merge

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_30536). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

Cyrilvallez commented:
@ArthurZucker With how things evolved in the end, the docstring for `DynamicCache` is still accurate!

ArthurZucker commented:
I am more thinking about explaining that we do internally some bucketing
but if you feel like it's good as is, we can merge

Cyrilvallez commented:
The bucketing part was removed in the end because it added unnecessary complexity to the code, and we could get all memory benefits without it! So I don't think that the docstring needs any more details (the `update()` method still work the same as before)

ArthurZucker commented:
Okay! Merging then! Congrats and great work 🤗 🚀
